File: server_akka_scala/src/main/scala/info/rori/lunchbox/server/akka/scala/domain/logic/PDFTextGroupStripper.scala
package info.rori.lunchbox.server.akka.scala.domain.logic

import org.apache.pdfbox.pdmodel.PDDocument
import org.apache.pdfbox.util.{TextPosition, PDFTextStripper}
import org.scalactic.Tolerance._
import org.scalactic.TripleEquals._

class PDFTextGroupStripper extends PDFTextStripper {
  var textGroups = Seq[TextGroup]()

  override def processTextPosition(text: TextPosition): Unit = {
    super.processTextPosition(text)

    textGroups.find(e => (e.xMax === text.getX +- 1.0f) && (e.yMin === text.getY +- 1.0f)) match {
      case Some(e) => e.add(text)
      case None => textGroups :+= TextGroup(text :: Nil)
    }
  }

  def getTextGroups(pDDocument: PDDocument): Seq[TextGroup] = {
    getText(pDDocument)
    textGroups
  }

  def getTextLines(pDDocument: PDDocument): Seq[TextLine] = TextLine.toLines(getTextGroups(pDDocument))
}

case class TextGroup(var positions: Seq[TextPosition]) {
  validate()

  def add(position: TextPosition): Unit = {
    validate()
    positions :+= position
  }

  def xMin: Float = positions.map(e => e.getX).min

  def xMax: Float = positions.map(e => e.getX + e.getWidth).max

  def yMin: Float = positions.map(e => e.getY).min

  def yMax: Float = positions.map(e => e.getY + e.getHeight).max

  def xMid: Float = (xMin + xMax) / 2

  def yMid: Float = (yMin + yMax) / 2

  def width: Float = xMax - xMin

  def height: Float = yMax - yMin

  def xIn(xVal: Float): Boolean = xMin <= xVal && xVal <= xMax

  def yIn(yVal: Float): Boolean = yMin <= yVal && yVal <= yMax

  override def toString: String = positions.map(_.getCharacter).mkString("").trim

  private def validate(): Unit = {
    require(positions.nonEmpty)
    require(positions.forall(positions.head.getY === _.getY +- 1.0f))
  }
}

case class TextLine(y: Float, texts: Seq[TextGroup]) {
  def oneTextMatches(regex: String): Boolean = texts.exists(_.toString.matches(regex))

  def allTextsMatch(regex: String): Boolean = texts.forall(_.toString.matches(regex))

  override def toString = texts.mkString(" ")
}

case object TextLine {

  def toLines(texts: Seq[TextGroup]): Seq[TextLine] = {
    var result = Map[Float, Seq[TextGroup]]()
    for (text <- texts) {
      result.keys.find(text.yIn) match {
        case Some(key) => result = result.updated(key, result(key) :+ text)
        case None => result += text.yMid -> Seq(text)
      }
    }
    result.toList.map(e => TextLine(e._1, e._2.sortBy(_.xMin))).sortBy(_.y)
  }
}

File: src/main/kotlin/lunchbox/util/pdf/PdfTextGroupStripper.kt
package lunchbox.util.pdf

import org.apache.pdfbox.pdmodel.PDDocument
import org.apache.pdfbox.text.PDFTextStripper
import org.apache.pdfbox.text.TextPosition
import kotlin.math.abs

fun Float.nearby(that: Float): Boolean = abs(this - that) <= 1.0f

class PdfTextGroupStripper : PDFTextStripper() {
  private val textGroups = mutableListOf<TextGroup>()

  override fun processTextPosition(text: TextPosition) {
    super.processTextPosition(text)

    val preText =
      textGroups.find { it.xMax().nearby(text.x) && it.yMin().nearby(text.y) }

    if (preText == null)
      textGroups += TextGroup(listOf(text))
    else
      preText.add(text)
  }

  fun getTextGroups(pDDocument: PDDocument): List<TextGroup> {
    getText(pDDocument)
    return textGroups
  }

  fun getTextLines(pDDocument: PDDocument): List<TextLine> =
    TextLine.toLines(getTextGroups(pDDocument))
}

data class TextGroup(var positions: List<TextPosition>) {
  init {
    validate()
  }

  fun add(position: TextPosition) {
    validate()
    positions += position
  }

  fun xMin(): Float = positions.map { it.x }.min() ?: 0f
  fun xMax(): Float = positions.map { it.x + it.width }.max() ?: 0f
  fun yMin(): Float = positions.map { it.y }.min() ?: 0f
  fun yMax(): Float = positions.map { it.y + it.height }.max() ?: 0f
  fun xMid(): Float = (xMin() + xMax()) / 2
  fun yMid(): Float = (yMin() + yMax()) / 2
  fun width(): Float = xMax() - xMin()
  fun height(): Float = yMax() - yMin()
  fun xIn(xVal: Float): Boolean = xMin() <= xVal && xVal <= xMax()
  fun yIn(yVal: Float): Boolean = yMin() <= yVal && yVal <= yMax()

  override fun toString(): String =
    positions.joinToString("").trim()

  private fun validate() {
    require(positions.isNotEmpty())
    require(positions.all { positions[0].y.nearby(it.y) })
  }
}

data class TextLine(val y: Float, val texts: List<TextGroup>) {
  fun oneTextMatches(regex: Regex): Boolean =
    texts.any { it.toString().matches(regex) }

  fun allTextsMatch(regex: Regex): Boolean =
    texts.all { it.toString().matches(regex) }

  override fun toString(): String =
    texts.joinToString(" ")

  companion object {
    fun toLines(texts: List<TextGroup>): List<TextLine> {
      val textMap = mutableMapOf<Float, List<TextGroup>>()
      for (text in texts) {
        val entry = textMap.entries.find { text.yIn(it.key) }
        if (entry == null)
          textMap[text.yMid()] = listOf(text)
        else
          entry.setValue(entry.value + text)
      }

      return textMap.entries
        .map { e -> TextLine(e.key, e.value.sortedBy { it.xMin() }) }
        .sortedBy { it.y }
    }
  }
}
